compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -501,7 +501,7 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
           ref(rhsClsSym.sourceModule)
         else
           inlineCallPrefix
-      val binding = ValDef(selfSym.asTerm, rhs).withSpan(selfSym.span).setDefTree
+      val binding = ValDef(selfSym.asTerm, rhs.changeOwner(ctx.owner,selfSym)).withSpan(selfSym.span).setDefTree
```

Comment:
This looks dangerous from a compilation-speed perspective. The invariant to keep is that the owners of definition inside the ValDef of symbol `x` must be either `x`, or one of `x`'s owners up to and including the method containing `x`. `changeOwner` is expensive since it does a full tree-traversal and it inspects every type in that tree. so we should do this only if it is absolutely necessary. 

One idea would be to do a changeOwner only if `!ctx.owner.isContainedIn(selfSym.enclosingMethodOrClass)`.

Reply:
Currently  -Y"check:macros" check more strict invariant:  that inside definition symbol should have the owner as nearest enclosing declaration:
 https://github.com/lampepfl/dotty/blob/a63b1d114179d79be32a1ee6552fc160c1bb2410/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala#L2804

 Maybe we should change this check, to keep a set of 'allowed owners'  in check traverse. 
On the other side, moving block of code from one method to another (as we do during inlining), requires changing owner, and with 'less strict invariant'  in any case full traverse will be needed (because we don't know - what in this tree).   With more strict (which enforced by current yCheck) - changeOwner can omit subtrees of low-level definitions since they already have correct owners.

 Don't know, what policy is better.

Reply:
Yes, quite possibly. I let @nicolasstucki comment on that.

Reply:
We have `QuoteUtils.changeOwnerOfTree` for this purpose. This variant checks if the tree has an owner that needs to be changed.

Reply:
Can we reproduce the failure without macros (only inline)? If not, this might be the wrong place to change the owner of that tree. It might also be an issue with the owners in the `TreeMap` or one of the methods the treemap calls.

Reply:
But yCheckOwner is defined in QuoteImpl,  so it will not be called without macros.
Note, that currently macros not change anything and used only to call yCheckOwner to tree.
(I can move yCheckOwner  be a separate check between phases and look).